### PR TITLE
Fixes #225 : Option to configure audit rules from this module itself

### LINKED ISSUE #225: Separate file for audit and option to configure audit rules from this module itself
Presently the module does not provide any option to configure audit rules from the module itself.
It would be really good if it would be possible to update audit rules from this module itself rather than
manually adding them or using a separate module for audit altogether.

In cases where Audit is used only with Wazuh (as in our case), this would be specifically useful as there would be no overhead of two modules, a single module would do it.

### PR DESCRIPTION
**Pull request by djmgit**
The PR implements the option to configure audit rules from the Wazuh module
it self. This also moves audit configuration and installation to a separate
pp file for better management.

### CODE REVIEW

**Comment by djmgit** [comment]
@rshad Please review.

**Comment by rshad** [comment]
Hi @djmgit !

Thank you for contributing to Wazuh, we really appreciate it.

I see you've done a great work here, it looks really interesting.

I'll review it during this week for sure. 

I'll be back soon.

Kr,

Rshad

**Comment by rshad** [comment]
Hi all!

Having set:
```
$audit_manage_rules                = true
```

The following template should be rendered.
https://github.com/wazuh/wazuh-puppet/blob/855de23d303a8cce5d973d243ed0228b82d9e507/templates/audit_rules.erb#L1-L18

As we checked, the template was rendered correctly.
- `/etc/audit/rules.d/audit.rules`
```bash
## First rule - delete all
-D

## Increase the buffers to survive stress events.
## Make this bigger for busy systems
-b 8192

## This determine how long to wait in burst of events
--backlog_wait_time 0

## Set failure mode to syslog
-f 1
```

**Related Puppet Log**
```
Info: Loading facts
Info: Caching catalog for agent-client-1
Info: Applying configuration version '1584538768'
Notice: /Stage[main]/Wazuh::Audit/Package[Installing Audit...]/ensure: created
Notice: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]/content:
--- /etc/audit/rules.d/audit.rules      2017-10-19 22:46:39.000000000 +0000
+++ /tmp/puppet-file20200318-2962-xpvkih        2020-03-18 13:39:55.954986245 +0000
@@ -1,14 +1,13 @@
-# This file contains the auditctl rules that are loaded
-# whenever the audit daemon is started via the initscripts.
-# The rules are simply the parameters that would be passed
-# to auditctl.
-
-# First rule - delete all
+## First rule - delete all
 -D

-# Increase the buffers to survive stress events.
-# Make this bigger for busy systems
--b 320
+## Increase the buffers to survive stress events.
+## Make this bigger for busy systems
+-b 8192
+
+## This determine how long to wait in burst of events
+--backlog_wait_time 0

-# Feel free to add below this line. See auditctl man page
+## Set failure mode to syslog
+-f 1


Info: Computing checksum on file /etc/audit/rules.d/audit.rules
Info: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]: Filebucketed /etc/audit/rules.d/audit.rules to puppet with sum f9869e1191838c461f5b9051c78a638d
Notice: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]/content: content changed '{md5}f9869e1191838c461f5b9051c78a638d' to '{md5}9967a3542c449b61d688a4f4b465cdfa'
Notice: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]/mode: mode changed '0640' to '0644'
Info: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]: Scheduling refresh of Service[auditd]
Info: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]: Scheduling refresh of Service[auditd]
Notice: /Stage[main]/Wazuh::Audit/Service[auditd]: Triggered 'refresh' from 2 events
```

---

**Adapted `audit_rules.pp` and the rules array**

Hi @djmgit,

I think having the template without already defined rules would be better, and these rules you consider that would come by default when enabling the rules, I moved them to the file `param_agent.pp`  to be defined in the array: 
```ruby
$audit_rules                       = []
```

As follows,
```pp
$audit_rules                       = [
        '-D',
        "-b ${audit_buffer_bytes}",
        "--backlog_wait_time ${audit_backlog_wait_time}",
        "-f 1"
]
```

Also, I adapted the template to be as follows,
```erb
<% if !@audit_rules.empty? -%>
<% @audit_rules.each do |audit_rule| -%>
<%= audit_rule %>
<%- end -%>
<%- end -%>
```

In this case, I re-installed the agent, and Puppet output is:
```bash
root@agent-client-1:~# puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for agent-client-1
Info: Applying configuration version '1584544019'
Notice: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]/content:
--- /etc/audit/rules.d/audit.rules      2020-03-18 15:05:49.611350290 +0000
+++ /tmp/puppet-file20200318-17015-ggysa4       2020-03-18 15:07:10.183616289 +0000
@@ -0,0 +1,5 @@
+-D
+-b 8192
+--backlog_wait_time 0
+-f 1
+

Info: Computing checksum on file /etc/audit/rules.d/audit.rules
Info: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]: Filebucketed /etc/audit/rules.d/audit.rules to puppet with sum d41d8cd98f00b204e9800998ecf8427e
Notice: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]/content:

Notice: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]/content: content changed '{md5}d41d8cd98f00b204e9800998ecf8427e' to '{md5}a11120ea77870ba2c05b40db0b7fc0e1'
Info: /Stage[main]/Wazuh::Audit/File[Configure audit.rules]: Scheduling refresh of Service[auditd]
Notice: /Stage[main]/Wazuh::Audit/Service[auditd]: Triggered 'refresh' from 1 events
Notice: Applied catalog in 1.41 seconds
```

- *`/etc/audit/rules.d/audit.rules`*
```
-D
-b 8192
--backlog_wait_time 0
-f 1
```

Kr,

Rshad

File: manifests/params_agent.pp [hunk 1]
```diff
@@ -224,6 +224,12 @@
       $ossec_syscheck_nodiff = '/etc/ssl/private.key'
       $ossec_syscheck_skip_nfs = 'yes'
 
+      # Audit
+      $audit_manage_rules                = false
+      $audit_buffer_bytes                = "8192"
+      $audit_backlog_wait_time           = "0"
+      $audit_rules                       = []
```
Comment (rshad) [inline manifests/params_agent.pp]: I would prefer to define the default rules you put in the template into this array, audit_rules.

I already changed it in 5972eee

Thanks,

File: templates/audit_rules.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+## Increase the buffers to survive stress events.
+## Make this bigger for busy systems
+-b <%= @audit_buffer_bytes %>
+
+## This determine how long to wait in burst of events
+--backlog_wait_time <%= @audit_backlog_wait_time %>
+
+## Set failure mode to syslog
+-f 1
+
```
Comment (rshad) [inline templates/audit_rules.erb]: As I commented in the previous comment, these rules should be defined in `params_agent.pp`.

I changed it in 5972eeef87dc906a072f41ff1714858d694c8498

Thanks.

File: manifests/params_agent.pp [hunk 3]
```diff
@@ -224,6 +224,17 @@
       $ossec_syscheck_nodiff = '/etc/ssl/private.key'
       $ossec_syscheck_skip_nfs = 'yes'
 
+      # Audit
+      $audit_manage_rules                = false
+      $audit_buffer_bytes                = "8192"
+      $audit_backlog_wait_time           = "0"
+      $audit_rules                       = [
+        '-D',
```
Comment (rshad) [inline manifests/params_agent.pp]: I also remove -D to not delete the already loaded audit rules.

Done in 16619a41df65abd9f1df59dfc262bd2ba7cde9a9

Thank you

**Comment by djmgit** [comment]
@rshad HI, the changes you made are very much reasonable.
The only thing that might be pointed out here is, now when a user of this module would want to add adudit rules from this puppet config, he/she would always have to precede his rules in the list by:
        "-b audit_buffer_bytes",
        "--backlog_wait_time audit_backlog_wait_time",
        "-f 1",
since the user's list will override the default rule list and also since there will be no default value for these options in the rules file.

**Comment by rshad** [comment]
> @rshad HI, the changes you made are very much reasonable.
> The only thing that might be pointed out here is, now when a user of this module would want to add audit rules from this puppet config, he/she would always have to precede his rules in the list by:
> "-b audit_buffer_bytes",
> "--backlog_wait_time audit_backlog_wait_time",
> "-f 1",
> since the user's list will override the default rule list and also since there will be no default value for these options in the rules file.

Hi @djmgit !

You're right. I would think about adding control variables to enable or not each one of the 3 rules we consider as default 
```
"-b audit_buffer_bytes",
"--backlog_wait_time audit_backlog_wait_time",
"-f 1",
```
In this case, the user would have more control about what to add or not. Also, I will work to make the configuration add rules to the already existing ones without overwriting the original rules file.

I will try to get it today.

Kr,

Rshad

**Comment by rshad** [comment]
Hi @djmgit !

We decided to remove the template for `audit` rules and changed from overwriting to appending to the rules file.

https://github.com/wazuh/wazuh-puppet/blob/697bad3e571aca7820308669e71f6cd52d67cac1/manifests/audit.pp#L33-L39

**Testing Log**

```
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for agent-client-1
Info: Applying configuration version '1584576544'
Notice: /Stage[main]/Wazuh::Audit/File_line[Append rule -b 8192 to /etc/audit/rules.d/audit.rules]/ensure: created
Notice: /Stage[main]/Wazuh::Audit/File_line[Append rule --backlog_wait_time 0 to /etc/audit/rules.d/audit.rules]/ensure: created
Notice: /Stage[main]/Wazuh::Audit/File_line[Append rule -f 1 to /etc/audit/rules.d/audit.rules]/ensure: created
Notice: Applied catalog in 1.08 seconds

```

Kr,

Rshad

**Pull request merged**